### PR TITLE
Protobuf support

### PR DESCRIPTION
**Pull request by Kingmidas74**
# Description

Provider for protobuf protocol was added as experimental feature. Some integration tests and acceptance tests is required, I think.

### CODE REVIEW

File: src/NClient/NClient.Common/Helpers/Converters.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+using System;
+
+namespace NClient.Common.Helpers
+{
+    public static class Converters
```
Comment (smolchanovsky) [inline src/NClient/NClient.Common/Helpers/Converters.cs:5]: This class must be internal

**Review by smolchanovsky — CHANGES_REQUESTED** [review]
The project `Nclient.Packages.Tests` contains tests that are run after the release of NuGet packages. They check that all packages are published and can be installed. You need to add a test for this package there

File: src/NClient/NClient.Common/Helpers/Converters.cs (hunk 1 above)
Comment (Kingmidas74) [reply src/NClient/NClient.Common/Helpers/Converters.cs:5]: Disagree. Can you explain? I think we can provide this solution of converts for user, aren't I right?

File: NClient.sln [hunk 2]
```diff
@@ -93,6 +93,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NClient.Providers.Api.Rest"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NClient.Providers.Api.Rest.Tests", "tests\NClient.Providers\NClient.Providers.Api.Rest.Tests\NClient.Providers.Api.Rest.Tests.csproj", "{27521108-9C67-44AE-885C-3BE343421955}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NClient.Providers.Serialization.Protobuf", "src\NClient.Providers\NClient.Providers.Serialization.Protobuf\NClient.Providers.Serialization.Protobuf.csproj", "{EC0A387A-AAEE-44EF-8AB0-6EF2D745E772}"
```
[the remark this answers is not in the document]
Comment (Kingmidas74) [reply NClient.sln]: Update name for naming convention. See new version of file.

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf/Extensions/ProtobufSerializationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+using NClient.Common.Helpers;
+
+namespace NClient.Providers.Serialization.Protobuf.Extensions
+{
+    public static class ProtobufSerializationExtensions
+    {
+        public static INClientOptionalBuilder<TClient, TRequest, TResponse> WithNewtonsoftJsonSerialization<TClient, TRequest, TResponse>(
```
[the remark this answers is not in the document]
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf/Extensions/ProtobufSerializationExtensions.cs]: Yep. My mistake. Ctrl+C -> Ctrl+V Driven Development :-( See new version of file.

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf/Extensions/ProtobufSerializationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+using NClient.Common.Helpers;
+
+namespace NClient.Providers.Serialization.Protobuf.Extensions
+{
+    public static class ProtobufSerializationExtensions
```
[the remark this answers is not in the document]
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf/Extensions/ProtobufSerializationExtensions.cs]: Add summaries. See new version of file.

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf/Extensions/UsingProtobufSerializationExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        public static INClientFactoryOptionalBuilder<TRequest, TResponse> UsingNewtonsoftJsonSerialization<TRequest, TResponse>(
+            this INClientFactorySerializationBuilder<TRequest, TResponse> serializationBuilder)
+        {
+            Ensure.IsNotNull(serializationBuilder, nameof(serializationBuilder));
+
+            return serializationBuilder.UsingCustomSerializer(new ProtobufSerializerProvider());
+        }
+
+        /// <summary>
+        /// Sets Newtonsoft.Json based <see cref="ISerializerProvider"/> used to create instance of <see cref="ISerializer"/>.
```
[the remark this answers is not in the document]
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf/Extensions/UsingProtobufSerializationExtensions.cs]: Yep. My mistake. Ctrl+C -> Ctrl+V Driven Development :-( See new version of file.

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf/NClient.Providers.Serialization.Protobuf.csproj [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    <PropertyGroup>
+        <PackageId>$(AssemblyName)</PackageId>
+        <VersionPrefix>$(VersionPrefix)</VersionPrefix>
+        <VersionSuffix>$(VersionSuffix)</VersionSuffix>
+        <PackageIcon>logo.png</PackageIcon>
+        <PackageProjectUrl>https://nclient.github.io</PackageProjectUrl>
+        <RepositoryUrl>https://github.com/nclient/nclient</RepositoryUrl>
+        <RepositoryType>git</RepositoryType>
+        <PackageTags>NClient;NClient.Providers;NClient.Providers.Serialization;Protobuf;Google</PackageTags>
```
[the remark this answers is not in the document]
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf/NClient.Providers.Serialization.Protobuf.csproj]: Add tag See new version of file.

File: src/NClient/NClient.Common/Helpers/Converters.cs (hunk 1 above)
Comment (smolchanovsky) [reply src/NClient/NClient.Common/Helpers/Converters.cs:5]: `NClient.Common` is not a package, so this converter is available only via the `NClient.Providers.Serialization.Protobuf.ProtobufNet` package. Why do users of this package need a converter?

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/UsingProtobufSerializationExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
+using NClient.Common.Helpers;
+
+namespace NClient.Providers.Serialization.Protobuf.ProtobufNet.Extensions
```
Comment (smolchanovsky) [inline src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/UsingProtobufSerializationExtensions.cs:3]: Extensions for creating a client are better defined in the namespace in which the client's builder is defined, like this:
```
// ReSharper disable once CheckNamespace
namespace NClient
```
It is needed for IDE to suggest this method together with the builder methods

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/ProtobufSerializationExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        /// <summary>
+        /// Sets ProtobufNet based <see cref="ISerializerProvider"/> used to create instance of <see cref="ISerializer"/>.
+        /// </summary>
+        /// <param name="optionalBuilder"></param>
+        /// <typeparam name="TClient"></typeparam>
+        /// <typeparam name="TRequest"></typeparam>
+        /// <typeparam name="TResponse"></typeparam>
+        /// <returns></returns>
+        public static INClientOptionalBuilder<TClient, TClient, TResponse> WithProtobufNetSerialization<TClient, TRequest, TResponse>(
```
Comment (smolchanovsky) [inline src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/ProtobufSerializationExtensions.cs:15]: The second parameter TClient must be replaced with the TRequest. This needs to be done in other methods too

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/UsingProtobufSerializationExtensions.cs (hunk 7 above)
Comment (smolchanovsky) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/UsingProtobufSerializationExtensions.cs:3]: `ProtobufSerializationExtensions` should also be defined in the `NClient` namespace

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/ProtobufSerializationExtensions.cs (hunk 8 above)
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/ProtobufSerializationExtensions.cs]: Fixed.

File: src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/UsingProtobufSerializationExtensions.cs (hunk 7 above)
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Serialization.Protobuf.ProtobufNet/Extensions/UsingProtobufSerializationExtensions.cs]: Fixed

**Comment by Kingmidas74** [comment]
I've fixed last mistakes. Sorry for the delay.

File: src/NClient/NClient.Common/Helpers/Converters.cs (hunk 1 above)
Comment (Kingmidas74) [reply src/NClient/NClient.Common/Helpers/Converters.cs]: Ok

**Review by smolchanovsky — APPROVED** [review]

**Comment by smolchanovsky** [comment]
LGTM

**Pull request merged**